apollo-biz/src/main/java/com/ctrip/framework/apollo/biz/service/ItemService.java

```diff
@@ -200,6 +201,18 @@ public Item update(Item item) {
 
   private boolean checkItemValueLength(long namespaceId, String value) {
     int limit = getItemValueLengthLimit(namespaceId);
+    Namespace currentNamespace = namespaceService.findOne(namespaceId);
+    if(currentNamespace != null) {
+        Matcher m = clusterPattern.matcher(currentNamespace.getClusterName());
+        boolean isGray = m.matches();
+        if (isGray) {
+          Namespace parentNamespace = namespaceService.findParentNamespace(currentNamespace);
```

Comment:
It looks quite clean now. BTW, shall we also check whether the parentNamespace is null here? e.g. some user creates the cluster with the name pattern as `[0-9]{14}-[a-zA-Z0-9]{16}`. And I suggest we split the logic into a different method, e.g.

```java
  private boolean checkItemValueLength(long namespaceId, String value) {
    int limit = getItemValueLengthLimit(namespaceId);
    Namespace currentNamespace = namespaceService.findOne(namespaceId);
    if(currentNamespace != null) {
        Matcher m = clusterPattern.matcher(currentNamespace.getClusterName());
        boolean isGray = m.matches();
        if (isGray) {
          limit = getGrayNamespaceItemValueLengthLimit(currentNamespace, limit);
        }
    }
    if (!StringUtils.isEmpty(value) && value.length() > limit) {
      throw new BadRequestException("value too long. length limit:" + limit);
    }
    return true;
  }

  private int getGrayNamespaceItemValueLengthLimit(Namespace grayNamespace, int grayNamespaceLimit) {
    Namespace parentNamespace = namespaceService.findParentNamespace(grayNamespace);
    if (parentNamespace != null) {
      int parentLimit = getItemValueLengthLimit(parentNamespace.getId());
      if (parentLimit > grayNamespaceLimit) {
        return parentLimit;
      }
    }
    return grayNamespaceLimit;
  }
```